pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,384 @@
+func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.RuntimeService, eventChannel chan *PodLifecycleEvent,
+	cache kubecontainer.Cache, genericPleg PodLifecycleEventGenerator, eventedPlegMaxStreamRetries int,
+	genericPlegRelistPeriod time.Duration, genericPlegRelistThreshold time.Duration, clock clock.Clock) PodLifecycleEventGenerator {
+	return &EventedPLEG{
+		runtime:                     runtime,
+		runtimeService:              runtimeService,
+		eventChannel:                eventChannel,
+		cache:                       cache,
+		genericPleg:                 genericPleg,
+		eventedPlegMaxStreamRetries: eventedPlegMaxStreamRetries,
+		genericPlegRelistPeriod:     genericPlegRelistPeriod,
+		genericPlegRelistThreshold:  genericPlegRelistThreshold,
+		clock:                       clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent events.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Relist relists all containers using GenericPLEG
+func (e *EventedPLEG) Relist() {
+	e.genericPleg.Relist()
+}
+
+// Start starts the Evented PLEG
+func (e *EventedPLEG) Start() {
+	e.runningMu.Lock()
+	defer e.runningMu.Unlock()
+	if !isEventedPLEGInUse {
+		isEventedPLEGInUse = true
+		e.stopCh = make(chan struct{})
+		e.stopCacheUpdateCh = make(chan struct{})
+		go wait.Until(e.watchEventsChannel, 0, e.stopCh)
+		go wait.Until(e.updateGlobalCache, 5*time.Second, e.stopCacheUpdateCh)
+	}
+}
+
+// Stop stops the Evented PLEG
+func (e *EventedPLEG) Stop() {
+	e.runningMu.Lock()
+	defer e.runningMu.Unlock()
+	if isEventedPLEGInUse {
+		close(e.stopCh)
+		close(e.stopCacheUpdateCh)
+		isEventedPLEGInUse = false
+	}
+}
+
+// In case the Evented PLEG experiences undetectable issues in the underlying
+// GRPC connection there is a remote chance the pod might get stuck in a
+// given state while it has progressed in it's life cycle. This function will be
+// called periodically to update the global timestamp of the cache so that those
+// pods stuck at GetNewerThan in pod workers will get unstuck.
+func (e *EventedPLEG) updateGlobalCache() {
+	e.cache.UpdateTime(time.Now())
+}
+
+// Update the relisting period and threshold
+func (e *EventedPLEG) Update(relistingPeriod time.Duration, relistThreshold time.Duration) {
+	e.genericPleg.Update(relistingPeriod, relistThreshold)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) >= cap(e.eventChannel) {
```

Comment:
Shouldn't this be only `==` ? i.e. len can't exceed the capacity.

Reply:
Yeah, right. Using `==` instead of `>=` now.